### PR TITLE
Implement single-shot semantics

### PR DESCRIPTION
**Pull request by rossberg**
Also rename `guard` to `barrier`.

### CODE REVIEW

File: interpreter/exec/eval.ml [hunk 1]
```diff
@@ -325,22 +325,30 @@ let rec step (c : config) : config =
       | Resume xls, Ref (NullRef _) :: vs ->
         vs, [Trapping "null continuation reference" @@ e.at]
 
-      | Resume xls, Ref (ContRef (n, ctxt)) :: vs ->
+      | Resume xls, Ref (ContRef {contents = None}) :: vs ->
+        vs, [Trapping "continuation resumed twice" @@ e.at]
```
Comment (dhil) [inline interpreter/exec/eval.ml:329]: Just a thought: in the future, we may want to make "multiple resume" and "null continuation" traps catchable exceptions.

File: interpreter/syntax/ast.ml [hunk 2]
```diff
@@ -100,7 +100,7 @@ and instr' =
   | Suspend of idx                    (* suspend continuation *)
   | Resume of (idx * idx) list        (* resume continuation *)
   | ResumeThrow of idx                (* abort continuation *)
-  | Guard of block_type * instr list  (* guard against suspension *)
+  | Barrier of block_type * instr list  (* guard against suspension *)
```
Comment (dhil) [inline interpreter/syntax/ast.ml:103]: Barriers, are they similar in spirit to Racket's [continuation barriers](https://docs.racket-lang.org/reference/eval-model.html#%28tech._continuation._barrier%29)?

Comment (rossberg) [reply interpreter/syntax/ast.ml:103]: Yes, they should be exactly that. Trying to capture one in a continuation leads to a trap.

File: interpreter/exec/eval.ml (hunk 1 above)
Comment (rossberg) [reply interpreter/exec/eval.ml:329]: No, I don't think so. These are essentially symptoms of codegen bugs in the producer, so a fatal failure ought to be adequate, just as for out-of-bounds memory access. If the producer really legally expects a null somewhere, then it should perform an explicit check.

**Pull request merged**